_posts/2020-07-21-cube-functionality-to-kartothek.markdown

```diff
@@ -0,0 +1,414 @@
+>>> print(", ".join(sorted(ds_latlong.indices)))
+country
+```
+Note that for the second dataset, no indices for **city** and **day** exists. 
+These are only created for the seed dataset, since that datasets forms the groundtruth about which city-day entries are part of the cube.
+(Dataset that provides the groundtruth about which Cells are in a Cube is called the **seed dataset**).
+
+If you look at the file tree, you can see that the second dataset is completely separated. This is useful to copy/backup parts of the cube:
+
+```python
+>>> print_filetree(store)
+geodata++latlong.by-dataset-metadata.json
+geodata++latlong/table/_common_metadata
+geodata++latlong/table/country=DE/<uuid>.parquet
+geodata++latlong/table/country=JP/<uuid>.parquet
+geodata++latlong/table/country=UK/<uuid>.parquet
+geodata++seed.by-dataset-metadata.json
+geodata++seed/indices/city/<ts>.by-dataset-index.parquet
+geodata++seed/indices/day/<ts>.by-dataset-index.parquet
+geodata++seed/table/_common_metadata
+geodata++seed/table/country=DE/<uuid>.parquet
+geodata++seed/table/country=UK/<uuid>.parquet
+```
+## Query
+
+The whole beauty of Cube does not come from storing multiple datasets,
+but especially from retrieving the data  (**Querying**)  in a very comfortable way. 
+Kartothek views the whole cube as a large, virtual DataFrame.
+The seed dataset presents the groundtruth regarding rows, all other datasets are joined via a left join. 
+Cube naturally supports **partition-by** semantic, which is more helpful for distributed backends.
+
+ 
+ ```python
+>>> from kartothek.io.eager_cube import query_cube
+>>> dfs = query_cube(
+...     cube=cube,
+...     store=store,
+...     partition_by="country",
+... )
+>>> dfs[0]
+   avg_temp     city country        day   latitude  longitude
+0         8  Dresden      DE 2020-07-01  51.050407  13.737262
+1         4  Dresden      DE 2020-07-02  51.050407  13.737262
+2         6  Hamburg      DE 2020-07-01  53.551086   9.993682
+3         5  Hamburg      DE 2020-07-02  53.551086   9.993682
+>>> dfs[1]
+   avg_temp    city country        day   latitude  longitude
+0         6  London      UK 2020-07-01  51.509865  -0.118092
+1         8  London      UK 2020-07-02  51.509865  -0.118092
+```
+
+The query system also supports selection and projection:
+
+```python
+>>> from kartothek.core.cube.conditions import C
+>>> from kartothek.io.eager_cube import query_cube
+>>> query_cube(
+...     cube=cube,
+...     store=store,
+...     payload_columns=["avg_temp"],
+...     conditions=(
+...         (C("country") == "DE") &
+...         C("latitude").in_interval(50., 52.) &
+...         C("longitude").in_interval(13., 14.)
+...     ),
+... )[0]
+   avg_temp     city country        day
+0         8  Dresden      DE 2020-07-01
+1         4  Dresden      DE 2020-07-02
+```
+
+ 
+## Transform
+
+Query and Extend can be combined to build powerful transformation pipelines. To better illustrate this we will use **dask.bag_cube** for that example.
+
```

Comment:
Is this really a feature of kartothek? We effectively only load data, what follows is dask. That topic feels like it would suit into another blog post, e.g. "how can you use kartothek to power ETL pipelines" or something more snappy

Reply:
The query planner and the modular split-map-reduce schema was one of the fundamental design goals of the cube functionality.